File: fsharp/pov/Pov.fs
module Pov

type Graph<'T> = { // tests nail down this naming, could introduce member functions instead, but well...
    value: 'T 
    children: Graph<'T> list }

type Path<'T> = Path of Value : 'T * LeftOfFocus : Graph<'T> list * RightOfFocus : Graph<'T> list 

type Zipper<'T> = Cursor of Focus : Graph<'T> * Paths : Path<'T> list

let mkGraph v ch = { value = v; children = ch }

let zipperFrom graph = Cursor(graph, []) 

let pathVal (Path(Value = v)) = v 

let traceFromZipper (Cursor(Focus = tr; Paths = paths )) =
    (List.map pathVal paths |> List.rev) @ [tr.value]

let down (Cursor(Focus = { value = v; children = chs }; Paths = paths)) = 
    match chs with // set cursor to first child (if existing) and prepend according path to paths
    | ch :: rs -> Cursor(ch, Path(v, [], rs) :: paths) |> Some 
    | _        -> None 

let right (Cursor(Focus = curr; Paths = paths)) = 
    match paths with // set cursor to next right child of the first path in paths list (if existing) and update this path
    | Path(v, lft, r :: rs) :: ps -> Cursor(r, Path(v, lft @ [curr], rs) :: ps) |> Some 
    | _                           -> None

let rec find v zipper = 
    let Cursor(Focus = f) = zipper
    if f.value = v then Some zipper 
    else 
        match down zipper |> Option.bind (find v) with 
        | Some x -> Some x 
        | None   -> right zipper |> Option.bind (find v) 

let rec reparent (Cursor(Focus = curr; Paths = paths)) = 
    match paths with 
    | []                      -> curr 
    | Path(v, lft, rgt) :: ps -> 
        let parentPerspective = reparent (Cursor(mkGraph v (lft @ rgt), ps))
        let newChildren = curr.children @ [parentPerspective]
        mkGraph curr.value newChildren

let fromPOV (v : 'T) (gr : Graph<'T>) : Graph<'T> option = 
    find v (zipperFrom gr) |> Option.map reparent

let tracePathBetween (v1 : 'T) (v2 : 'T) (gr : Graph<'T>) : 'T list option = 
    fromPOV v1 gr |> Option.bind (zipperFrom >> find v2 >> Option.map traceFromZipper)

File: elixir/pov/lib/pov.ex
defmodule Pov do
  @typedoc """
  A tree, which is made of a node with several branches
  """
  @type tree :: {any, [tree]}

  # see the solution in F#, in Elixir additional types
  # have to be defined in their own modules
  defmodule Path do
    defstruct [:value, :lft, :rgt] # left_of_focus and right_of_focus
    @type t :: %Path{value: any, lft: [Pov.tree()], rgt: [Pov.tree()]}
  end
  defmodule Zipper do
    defstruct [:focus, :paths]
    @type t :: %Zipper{focus: Pov.tree(), paths: [Path.t()]}
  end

  @doc """
  Reparent a tree on a selected node.
  """
  @spec from_pov(tree :: tree, node :: any) :: {:ok, tree} | {:error, atom}
  def from_pov(tree, node) do
    case zipper_from(tree) |> find(node) do
      {:ok, zipper} -> {:ok,    reparent(zipper)}
      _             -> {:error, :nonexistent_target}
    end
  end

  @doc """
  Finds a path between two nodes
  """
  @spec path_between(tree :: tree, from :: any, to :: any) :: {:ok, [any]} | {:error, atom}
  def path_between(tree, from, to) do
    case zipper_from(tree) |> find(from) do
      {:ok, zipper} ->
        case zipper_from(reparent(zipper)) |> find(to) do
          {:ok, nxt_zipper} -> {:ok,    trace_from_zipper(nxt_zipper)}
          _                 -> {:error, :nonexistent_destination}
        end
      _            -> {:error, :nonexistent_source}
    end
  end
  # private helper methods
  defp zipper_from(tree), do: %Zipper{focus: tree, paths: []}
  defp path_value(%Path{value: v}), do: v
  defp trace_from_zipper(%Zipper{focus: {v, _}, paths: paths}) do
    [v | Enum.map(paths,&path_value/1)] |> Enum.reverse
  end
  # set cursor to first child (if existing) and prepend according path to paths
  defp down(%Zipper{focus: {v, [ch | chs]}, paths: paths}) do
    %Zipper{focus: ch, paths: [%Path{value: v, rgt: chs, lft: []} | paths]}
  end
  defp down(_), do: nil # Elixir way of getting a None
  # set cursor to next right child of the first path in paths list (if existing) and update this path
  defp right(%Zipper{focus: tree, paths: [%Path{lft: lft, rgt: [rgt | rgts]} = path | ps]}) do
    %Zipper{focus: rgt, paths: [%Path{path | lft: [tree | lft], rgt: rgts} | ps]}
  end
  defp right(_), do: nil
  defp find(%Zipper{focus: {node, _}} = zipper, node), do: {:ok, zipper}
  defp find(nil, _), do: nil
  defp find(%Zipper{} = zipper, node) do
    case down(zipper) |> find(node) do # recurse at next level and match
      {:ok, z} -> {:ok, z}
      _        -> right(zipper) |> find(node)
    end
  end
  defp reparent(%Zipper{focus: tree, paths: []}), do: tree
  defp reparent(%Zipper{focus: {focus_v, chs}, paths: [%Path{value: path_v, lft: lft, rgt: rgt} | ps]}) do
    parent_perspective = reparent(%Zipper{focus: {path_v, lft ++ rgt}, paths: ps})
    {focus_v, [parent_perspective | chs]} # add the parent perspective to the children
  end
end
